<a href="https://colab.research.google.com/github/jjgallonc/Analitica/blob/main/Reto2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Cargar Librerías de referencia

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import LabelEncoder

1. Se transforman las variables categoricas

In [ ]:
le=LabelEncoder()
XD[:,0]=le.fit_transform(XD[:,0])
YD=le.fit_transform(YD)

/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:114: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


2. Cargar datos del archivo

In [ ]:
nxl='/content/DB.Reto2.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XDB=XDB.dropna()

XD=np.array(XDB[['College','FoodSpend','Income','TravelSpend']])
YD=np.array(XDB[['CreditCard']])

3. Creación del modelo

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD,YD)
ydp=mar.predict(XD)

from sklearn.metrics import confusion_matrix
cm=confusion_matrix(YD,ydp)
print(cm)

VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]

Exactitud=(VP+VN)/(VP+VN+FP+FN)
Sensibilidad=VP/(VP+FN)
Especificidad=VN/(VN+FP)

print('La Exactitud del Modelo fue:', Exactitud)
print('La Sensibilidad del Modelo fue:', Sensibilidad)
print('La Especificidad del Modelo fue:', Especificidad)

[[334   9]
 [ 42 115]]
La Exactitud del Modelo fue: 0.898
La Sensibilidad del Modelo fue: 0.732484076433121
La Especificidad del Modelo fue: 0.9737609329446064


4. Cargar librerías para graficar el arbol

In [ ]:
from sklearn.tree import export_graphviz
from pydotplus import graph_from_dot_data

vs=['College','FoodSpend','Income','TravelSpend']
dot_graph=export_graphviz(mar,feature_names=vs)
graph=graph_from_dot_data(dot_graph)
graph.write_png('arbol.png')

True

5. Evaluar personas

In [ ]:
Xun1=np.array([[0,2892.9,65982,2151.6]])
ydun1=mar.predict(Xun1)
print(ydun1)

Xun2=np.array([[1,6017.66,53986,2002.5]])
ydun2=mar.predict(Xun2)
print(ydun2)

[0]
[0]


El modelo de árbol de decisión utiliza las variables "College", "Food Spend", "Income", y "Travel Spend" para predecir si un cliente usará su tarjeta de crédito en gastos de viaje y alimentos, siendo "Credit Card" la variable objetivo. Con una exactitud del 89.8%, el modelo demuestra un alto nivel de precisión general, clasificando correctamente la mayoría de los casos. La sensibilidad del 73.2% indica que es razonablemente efectivo en identificar a los clientes que usarán la tarjeta. Por otro lado, la especificidad del 97.4% muestra que el modelo es muy eficiente en evitar casos negativos, identificando correctamente a aquellos que no usarán la tarjeta. En cuanto a las reglas de negocio, para el individuo 1 se determinó que si el "Travel Spend" es mayor o igual a 1567 y "Income" supera los 59098, es probable que use la tarjeta; mientras que para el individuo 2, si el "Travel Spend" es mayor o igual a 1567, el "Income" es menor a 59098, y "Food Spend" es mayor a 2711, también se considera un posible usuario de crédito. En resumen, el modelo de árbol de decisión se presenta como una herramienta confiable y eficaz para predecir el uso de tarjetas de crédito, destacando en la identificación precisa de los casos negativos, con una buena capacidad para clasificar correctamente los positivos.